File: archive/cdk-go/Pr2360/lib/storage_stack.go
package lib

import (
	"github.com/aws/aws-cdk-go/awscdk/v2"
	"github.com/aws/aws-cdk-go/awscdk/v2/awsec2"
	"github.com/aws/aws-cdk-go/awscdk/v2/awselasticache"
	"github.com/aws/aws-cdk-go/awscdk/v2/awsrds"
	"github.com/aws/aws-cdk-go/awscdk/v2/awss3"
	"github.com/aws/constructs-go/constructs/v10"
	"github.com/aws/jsii-runtime-go"
)

type StorageStackProps struct {
	*awscdk.NestedStackProps
	EnvironmentSuffix     *string
	Vpc                   awsec2.IVpc
	DatabaseSecurityGroup awsec2.ISecurityGroup
}

type StorageStack struct {
	awscdk.NestedStack
	Database     awsrds.IDatabaseCluster
	S3Bucket     awss3.IBucket
	RedisCluster awselasticache.CfnCacheCluster
}

func NewStorageStack(scope constructs.Construct, id *string, props *StorageStackProps) *StorageStack {
	nestedStack := awscdk.NewNestedStack(scope, id, props.NestedStackProps)

	// Get environment suffix with fallback
	envSuffix := "dev"
	if props != nil && props.EnvironmentSuffix != nil && *props.EnvironmentSuffix != "" {
		envSuffix = *props.EnvironmentSuffix
	}

	// S3 Bucket for static assets
	bucket := awss3.NewBucket(nestedStack, jsii.String("AssetsBucket"), &awss3.BucketProps{
		BucketName:        jsii.String("tap-assets-" + envSuffix),
		Versioned:         jsii.Bool(true),
		RemovalPolicy:     awscdk.RemovalPolicy_DESTROY,
		AutoDeleteObjects: jsii.Bool(true),
		BlockPublicAccess: awss3.BlockPublicAccess_BLOCK_ALL(),
		Encryption:        awss3.BucketEncryption_S3_MANAGED,
		IntelligentTieringConfigurations: &[]*awss3.IntelligentTieringConfiguration{
			{
				Name:                      jsii.String("EntireBucket"),
				ArchiveAccessTierTime:     awscdk.Duration_Days(jsii.Number(90)),
				DeepArchiveAccessTierTime: awscdk.Duration_Days(jsii.Number(180)),
			},
		},
	})

	// Aurora Serverless v2 Database
	cluster := awsrds.NewDatabaseCluster(nestedStack, jsii.String("AuroraCluster"), &awsrds.DatabaseClusterProps{
		Engine: awsrds.DatabaseClusterEngine_AuroraMysql(&awsrds.AuroraMysqlClusterEngineProps{
			Version: awsrds.AuroraMysqlEngineVersion_VER_3_07_0(),
		}),
		Writer: awsrds.ClusterInstance_ServerlessV2(jsii.String("writer"), &awsrds.ServerlessV2ClusterInstanceProps{
			ScaleWithWriter: jsii.Bool(true),
		}),
		Readers: &[]awsrds.IClusterInstance{
			awsrds.ClusterInstance_ServerlessV2(jsii.String("reader"), &awsrds.ServerlessV2ClusterInstanceProps{
				ScaleWithWriter: jsii.Bool(false),
			}),
		},
		ServerlessV2MinCapacity: jsii.Number(0.5),
		ServerlessV2MaxCapacity: jsii.Number(16),
		Vpc:                     props.Vpc,
		VpcSubnets: &awsec2.SubnetSelection{
			SubnetType: awsec2.SubnetType_PRIVATE_ISOLATED,
		},
		SecurityGroups:      &[]awsec2.ISecurityGroup{props.DatabaseSecurityGroup},
		DefaultDatabaseName: jsii.String("tapdb"),
		RemovalPolicy:       awscdk.RemovalPolicy_DESTROY,
		DeletionProtection:  jsii.Bool(false),
	})

	// ElastiCache Redis Subnet Group
	subnetGroup := awselasticache.NewCfnSubnetGroup(nestedStack, jsii.String("RedisSubnetGroup"), &awselasticache.CfnSubnetGroupProps{
		Description: jsii.String("Subnet group for Redis cluster"),
		SubnetIds: props.Vpc.SelectSubnets(&awsec2.SubnetSelection{
			SubnetType: awsec2.SubnetType_PRIVATE_WITH_EGRESS,
		}).SubnetIds,
	})

	// ElastiCache Redis Cluster
	redisCluster := awselasticache.NewCfnCacheCluster(nestedStack, jsii.String("RedisCluster"), &awselasticache.CfnCacheClusterProps{
		CacheNodeType:        jsii.String("cache.t3.micro"),
		Engine:               jsii.String("redis"),
		NumCacheNodes:        jsii.Number(1),
		CacheSubnetGroupName: subnetGroup.Ref(),
		VpcSecurityGroupIds:  &[]*string{props.DatabaseSecurityGroup.SecurityGroupId()},
	})

	return &StorageStack{
		NestedStack:  nestedStack,
		Database:     cluster,
		S3Bucket:     bucket,
		RedisCluster: redisCluster,
	}
}

File: archive/cdk-ts/Pr3483/lib/storage-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as elasticache from 'aws-cdk-lib/aws-elasticache';
import * as opensearch from 'aws-cdk-lib/aws-opensearchservice';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

interface StorageStackProps {
  vpc: ec2.Vpc;
  environmentSuffix: string;
}

export class StorageStack extends Construct {
  public readonly redisCluster: elasticache.CfnCacheCluster;
  public readonly openSearchDomain: opensearch.Domain;
  public readonly mediaBucket: s3.Bucket;
  public readonly redisSecurityGroup: ec2.SecurityGroup;
  public readonly openSearchSecurityGroup: ec2.SecurityGroup;

  constructor(scope: Construct, id: string, props: StorageStackProps) {
    super(scope, id);

    // Redis Security Group
    this.redisSecurityGroup = new ec2.SecurityGroup(
      this,
      'RedisSecurityGroup',
      {
        vpc: props.vpc,
        description: 'Security group for ElastiCache Redis',
        allowAllOutbound: false,
      }
    );

    // Redis Subnet Group
    const redisSubnetGroup = new elasticache.CfnSubnetGroup(
      this,
      'RedisSubnetGroup',
      {
        description: 'Subnet group for Redis cache',
        subnetIds: props.vpc.selectSubnets({
          subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
        }).subnetIds,
      }
    );

    // ElastiCache Redis Cluster
    this.redisCluster = new elasticache.CfnCacheCluster(
      this,
      'WikiRedisCache',
      {
        cacheNodeType: 'cache.t3.micro',
        engine: 'redis',
        numCacheNodes: 1,
        cacheSubnetGroupName: redisSubnetGroup.ref,
        vpcSecurityGroupIds: [this.redisSecurityGroup.securityGroupId],
        preferredMaintenanceWindow: 'sun:05:00-sun:06:00',
        snapshotRetentionLimit: 7,
        snapshotWindow: '03:00-04:00',
        tags: [
          {
            key: 'Name',
            value: `WikiRedisCache-${props.environmentSuffix}`,
          },
          {
            key: 'Environment',
            value: props.environmentSuffix,
          },
        ],
      }
    );

    // OpenSearch Security Group
    this.openSearchSecurityGroup = new ec2.SecurityGroup(
      this,
      'OpenSearchSecurityGroup',
      {
        vpc: props.vpc,
        description: 'Security group for OpenSearch domain',
        allowAllOutbound: true,
      }
    );

    // OpenSearch Domain
    this.openSearchDomain = new opensearch.Domain(this, 'WikiSearchDomain', {
      version: opensearch.EngineVersion.OPENSEARCH_2_11,
      capacity: {
        dataNodeInstanceType: 't3.small.search',
        dataNodes: 2,
        multiAzWithStandbyEnabled: false,
      },
      ebs: {
        volumeSize: 20,
        volumeType: ec2.EbsDeviceVolumeType.GP3,
      },
      nodeToNodeEncryption: true,
      encryptionAtRest: {
        enabled: true,
      },
      vpcSubnets: [
        props.vpc.selectSubnets({
          subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
        }),
      ],
      securityGroups: [this.openSearchSecurityGroup],
      logging: {
        slowSearchLogEnabled: true,
        appLogEnabled: true,
        slowIndexLogEnabled: true,
      },
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    // S3 Bucket for Media Uploads
    this.mediaBucket = new s3.Bucket(this, 'WikiMediaBucket', {
      bucketName: `wiki-media-${props.environmentSuffix}-${cdk.Stack.of(this).account}`,
      versioned: true,
      encryption: s3.BucketEncryption.S3_MANAGED,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      lifecycleRules: [
        {
          id: 'DeleteOldVersions',
          noncurrentVersionExpiration: cdk.Duration.days(90),
          abortIncompleteMultipartUploadAfter: cdk.Duration.days(7),
        },
      ],
      cors: [
        {
          allowedMethods: [
            s3.HttpMethods.GET,
            s3.HttpMethods.POST,
            s3.HttpMethods.PUT,
          ],
          allowedOrigins: ['*'],
          allowedHeaders: ['*'],
          maxAge: 3000,
        },
      ],
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    // Tags
    cdk.Tags.of(this.mediaBucket).add(
      'Name',
      `WikiMediaBucket-${props.environmentSuffix}`
    );
    cdk.Tags.of(this.mediaBucket).add('Environment', props.environmentSuffix);
  }
}
